web/js/stats/stats.js

```diff
@@ -0,0 +1,348 @@
+     */
+    const moduleUi = (label = '', withGraph = false, postfix = 'ms') => {
+        const ui = document.createElement('div'),
+            _label = document.createElement('div'),
+            _value = document.createElement('span');
+        ui.append(_label, _value);
+
+        let postfix_ = postfix;
+
+        let _graph;
+        if (withGraph) {
+            const _container = document.createElement('span');
+            ui.append(_container);
+            _graph = graph(_container);
+        }
+
+        _label.innerHTML = label;
+
+        const withPostfix = (value) => postfix_ = value;
+
+        const update = (value) => {
+            if (_graph) _graph.add(value);
+            // 203 (333) ms
+            _value.textContent = `${value < 1 ? '<1' : value} ${_graph ? `(${_graph.max()}) ` : ''}${postfix_}`;
+        }
+
+        return {el: ui, update, withPostfix}
+    }
+
+    /**
+     * Latency stats submodule.
+     *
+     * Accumulates the simple rolling delta mean value
+     * between a server request and a following server response values.
+     *
+     *      window
+     *   _____________
+     *  |            |
+     * [1, 1, 3, 4, 1, 4, 3, 1, 2, 1, 1, 1, 2, ... n]
+     *              |
+     *    stats_snapshot_period
+     *    mean = round(next - mean / length % window)
+     *
+     * Events:
+     * <- PING_RESPONSE
+     * <- PING_REQUEST
+     *
+     * ?Interface:
+     *  HTMLElement get()
+     *  void enable()
+     *  void disable()
+     *  void render()
+     *
+     * @version 1
+     */
+    const latency = (() => {
+        let listeners = [];
+
+        let mean = 0;
+        let length = 0;
+        let previous = Date.now();
+        const window = 5;
+
+        // UI
+        const ui = moduleUi('Ping', true);
+
+        const onPingRequest = (data) => previous = data.time;
+
+        const onPingResponse = () => {
+            length++;
+            const delta = Date.now() - previous;
+            mean += Math.round((delta - mean) / length);
+
+            if (length % window === 0) {
+                length = 1;
+                mean = delta;
+            }
+        }
+
+        const enable = () => {
+            listeners.push(
+                event.sub(PING_RESPONSE, onPingResponse),
+                event.sub(PING_REQUEST, onPingRequest)
+            );
+        }
+
+        const disable = () => {
+            while (listeners.length) listeners.shift().unsub();
+        }
+
+        const render = () => ui.update(mean);
+
+        const get = () => ui.el;
+
+        return {get, enable, disable, render}
+    })(event, moduleUi);
+
+    /**
+     * User agent memory stats.
+     *
+     * ?Interface:
+     *  HTMLElement get()
+     *  void enable()
+     *  void disable()
+     *  void render()
+     *
+     * @version 1
+     */
+    const clientMemory = (() => {
+        let active = false;
+
+        const ui = moduleUi('Memory', false, 'B');
```

Comment:
I think you can write it more naturally by provide postfix as a format function with value input.
ui = moduleUi('Memory', false, formatMemory);
ui = moduleUi('Latency', false, formatLatency);

Reply:
Mmm... probably. I'll think.